core/charm/arches.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package charm
+
+import (
+	"sort"
+)
+
+// Arch represents a platform architecture.
+type Arch string
+
+func (a Arch) String() string {
+	return string(a)
+}
+
+const (
+	// ArchAll represents an architecture for all architectures.
+	ArchAll Arch = "all"
+
+	// ArchAMD64 defines a amd64 architecture.
+	ArchAMD64 Arch = "amd64"
+
+	// ArchARM64 defines a arm64 architecture.
+	ArchARM64 Arch = "arm64"
+
+	// ArchPPC64 defines a ppc64 architecture.
+	ArchPPC64 Arch = "ppc64"
+
+	// ArchS390 defines a s390 architecture.
+	ArchS390 Arch = "s390"
+)
+
+// Arches defines a list of arches to compare against.
+type Arches struct {
+	set map[Arch]struct{}
+}
+
+// DefaultArches creates a series of arches to compare against.
+func DefaultArches() Arches {
```

Comment:
I am not fond of having `ArchAll` in the map; it reads kinda hacky. 

What about renaming this method to `AllArches` that returns the map _without ArchAll_. Then you could change your set checks to something like `arch == charm.ArchAll || charm.AllArches().Contains(arch)`.